lib/cog/pipeline/relay_selector.ex

```diff
@@ -0,0 +1,52 @@
+defmodule Cog.Pipeline.RelaySelector do
+
+  alias Cog.Relay.Relays
+
+  @moduledoc ~s"""
+  Selects a Relay for a given bundle name and version
+  """
+
+  defstruct [:bundle_name, :bundle_version, :relay]
+
+  @doc "Creates a new RelaySelector"
+  @spec new(String.t, String.t) :: RelaySelector.t
+  def new(bundle_name, bundle_version) do
+    %__MODULE__{bundle_name: bundle_name, bundle_version: bundle_version}
+  end
+
+  @doc "Selects a Relay and updates the selector's internal state."
+  @spec select(RelaySelector.t) :: {:ok, RelaySelector.t} | {:error, any}
+  def select(%__MODULE__{bundle_name: name, bundle_version: version, relay: nil}=selector) do
+    case Relays.pick_one(name, version) do
+      {:ok, relay} ->
+        selector = %{selector | relay: relay}
+        {:ok, selector}
+      error ->
+      # Query DB to clarify error before reporting to the user
+      if Cog.Repository.Bundles.assigned_to_group?(name) do
+        error
+      else
+        {:error, {:no_relay_group, name}}
+      end
+    end
+  end
+  def select(%__MODULE__{bundle_name: name, bundle_version: version, relay: relay}=selector) do
+    if Relays.relay_available?(relay, name, version) do
+      {:ok, selector}
+    else
+      select(%{selector | relay: nil})
+    end
+  end
+
+  @doc "Returns the name of the selected Relay"
+  @spec relay(RelaySelector.t) :: String.t | nil
+  def relay(%__MODULE__{relay: relay}), do: relay
+
+  @doc "Constructs a MQTT topic for the named command"
+  @spec relay_topic(RelaySelector.t, String.t) :: String.t
+  def relay_topic(%__MODULE__{relay: nil}, _command_name), do: ""
```

Comment:
The sequence of using `RelaySelector` goes like this:

```
# Create a selector for a given bundle and version
selector = RelaySelector.new("my_bundle", "1.3.1")
# Ask the selector to select a relay
{:ok, selector} = RelaySelector.select(selector)
# Build the topic for the selected relay
topic = RelaySelector.relay_topic(selector)
```
If the caller creates a selector and calls `relay_topic/2` without first calling `select/1` then there's no relay from which to build a topic string.